ipc: add bitcoin-mine test program

ryanofsky opened the pull request:
Add [`bitcoin-mine`](https://github.com/ryanofsky/bitcoin/blob/pr/mine/src/bitcoin-mine.cpp#L20-L37) executable to test connecting to the `bitcoin-node` process over a unix socket and calling [`interface::Mining`](https://github.com/ryanofsky/bitcoin/blob/pr/mine/src/interfaces/mining.h) methods.

This is useful as discussed in https://github.com/bitcoin/bitcoin/pull/29432#issuecomment-2222418506 to work on getting mining code working in an external process.

This PR initially supported spawning a new `bitcoin-node` process if a connection could not be established with an existing process, but it was dropped to simplify the code. A branch building on this PR and adding that support is
[pr/mine-detach](https://github.com/ryanofsky/bitcoin/commits/pr/mine-detach).

---

This PR is part of the [process separation project](https://github.com/bitcoin/bitcoin/issues/28722).

ryanofsky committed:
interface_ui: move from node to common library

Allows InitError and InitWarning to be used by other executables beside
bitcoind and bitcoin-node.

Sjors commented:
Thanks. I'll try to get the Mining interface in good shape as well, which this can then be rebased on:

- #30409 (first commit)
- #30356
- to have `createNewBlock()` return an interface, see https://github.com/bitcoin/bitcoin/pull/29432#issuecomment-2225337100
- add `waitFeesChanged()` (initially just uses `GetTransactionsUpdated()`, but once cluster mempool is in place it can do frequent (fast) block template construction to see if fees have actually gone up.

ryanofsky commented:
I'll work on dropping the spawn functionality from this PR next, and just having the `bitcoin-mine` program print an error when it can't connect to the node socket, instead of starting the node itself. Sjors suggested this in https://github.com/bitcoin/bitcoin/pull/29432#issuecomment-2225041208:

> If it helps getting IPC stuff merged quicker, it's fine by me if the miner can't spin up a node process (see e.g. [Sjors@b6dfa5f](https://github.com/Sjors/bitcoin/commit/b6dfa5f186e8a4813db21b2547de98280643f51a))

This would simplify the PR by allowing 3 commits to be dropped:  

- [`a06a4fa77188` multiprocess: Add IPC spawnProcess detach option](https://github.com/bitcoin/bitcoin/commit/a06a4fa77188feae7b5731707f447ddaf0961898)
- [`b53604a1f547` multiprocess: Add IPC attach and detach methods](https://github.com/bitcoin/bitcoin/commit/b53604a1f5475a65476235bebecdf7ff44911ba7)
- [`cfa8f3a9a3cd` multiprocess: Add startNode and stopNode methods to Mining interface](https://github.com/bitcoin/bitcoin/commit/cfa8f3a9a3cdb26af03985d9358f7928bc7e7fec)

Sjors commented:
> filling in missing serialization for `CBlockTemplate`

You won't need that if you rebase on #30440, though maybe it's better to wait until that's merged to avoid too much code churn.

ryanofsky commented:
> > filling in missing serialization for `CBlockTemplate`
> 
> You won't need that if you include #30440, though maybe it's better to wait until that's merged to avoid too much code churn.

Yeah definitely some of that should not be needed. Some parts will probably be needed though. For example, the initial version of this PR did not support deserializing `CTransaction` or `CBlock` objects (because `CTransaction`  does not have an `Unserialize` method, and also requires a version parameter to be set).

Sjors commented:
I'll certainly need `CBlock`. It's probably good to support `CTransaction` as well, as it might allow for more efficiently fetching only missing transactions in a template update, but that would be a future optimisation.

Sjors commented:
I opened draft PRs for all interface changes that I think I will need, see updated comment above: https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2225416191

Sjors commented:
I wrote:

> I'll expand the demo app to simulate all the steps a Template Provider would do

Once that works the next step for me would be to open an alternative to #29432 which runs the Template Provider in `bitcoin-tp`. I might wait with that until we ship a release with this interface enabled on the node side, ideally in `bitcoind` and `bitcoin-qt` rather than in a seperate `bitcoin-node` and `bitcoin-gui` binaries.

I could then modify the guix build to _only_ produce `bitcoin-tp`. This effectively turns it into a separate project that just happens to have access to the entire Bitcoin Core codebase.

If later on we can also turn `libbitcoin_net` into a library (see #30350), I could look into completely separating the codebase (I'd probably copy-paste the Bitcoin Core code and strip everything out I don't need, rather starting a new c++ project and build system from scratch). E.g. it could pull in `libbitcoin_net`, `libbitcoin_util` and whatever else it needs, add its own `Transport` subclass. This lets me reuse #29346, #30315 and #30332, but without having to merge them into Bitcoin Core itself.

This would be in parallel to the alternative approach of writing `bitcoin-tp` from scratch in rust reusing SRI code.

ryanofsky commented:
I simplified PR as suggested so `bitcoin-mine` no longer spawns `bitcoin-node` if it can't connect to it. Instead it just prints an error message and suggests a command line to run `bitcoin-node` manually.

I also added a new commit to make the IPC code compatible with all the interfaces added in https://github.com/bitcoin/bitcoin/pull/30409, https://github.com/bitcoin/bitcoin/pull/30356, https://github.com/bitcoin/bitcoin/pull/30440, and https://github.com/bitcoin/bitcoin/pull/30443, so this PR is easier to test with those interfaces.

I still need to break up the main commit ("Add bitcoin-mine test program") and add unit tests for the new connecting, listening, and serialization code that's introduced.

Rebased 3c0a13c9a56a11d795b243cedb3b00a71a81042c -> 4e1a4342f3b2a857e3211587cd14e36704197483 ([`pr/mine.2`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.2) -> [`pr/mine.3`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.2-rebase..pr/mine.3)) with those changes.

For reference, the code that was dropped from this PR allowing bitcoin-mine to spawn a bitcoin-node process is in commits:

- 09621da38291fbfc22d86ea98e4499b3aa2c9976 multiprocess: Add IPC spawnProcess detach option
- 3d23332240f103f6c639662d7c732a52adf5d13f multiprocess: Add IPC attach and detach methods
- 3c112629d99d64b1c7719cd432d13fb2e8ec8f4b multiprocess: Allow bitcoin-mine to spawn bitcoin-node

These changes would probably be useful in another context, but aren't needed for this test program.

re: https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2226847354

I think your ideas for packaging bitcoin-tp make sense. And maybe it would also be good to try to integrate IPC in bitcoind and bitcoin-qt executables instead of separate bitcoin-node and bitcoin-gui executables. Would have to think about how to implement this and what the tradeoffs would be.

Sjors commented on src/bitcoin-mine.cpp:
```diff
@@ -0,0 +1,115 @@
+#include <compat/compat.h>
+#include <init/common.h>
+#include <interfaces/init.h>
+#include <interfaces/ipc.h>
```
8a20be71f07cfe0129107cb8effb2490b79639bf: `#include <logging.h>`

Sjors commented:
If you split the capnp additions out of 4e1a4342f3b2a857e3211587cd14e36704197483 it's a bit easier for me to combine this PR with #29432 (by omitting the interface changes commit here or there).

Sjors commented on src/interfaces/mining.h:
```diff
@@ -46,6 +72,9 @@ class Mining
     virtual std::unique_ptr<node::CBlockTemplate> createNewBlock(const CScript& script_pub_key, bool use_mempool = true) = 0;
 
+    // Implemented in https://github.com/bitcoin/bitcoin/pull/30356
+    virtual std::unique_ptr<BlockTemplate> createNewBlock2(const CScript& script_pub_key, const node::BlockCreateOptions& options={}) { return {}; }
```
Is it possible to rename this to `createNewBlock` and drop the one above? Or is there an inconsistency between my PRs?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30437#discussion_r1681055963

> Is it possible to rename this to `createNewBlock` and drop the one above? Or is there an inconsistency between my PRs?

It's not really possible without making the PR bigger because mining code relies on createNewBlock returning CBlockTemplate not BlockTemplate, and I want mining code to work and CI to pass. 

The purpose of createNewBlock2 is just to make sure serialization works when createNewBlock is changed to return BlockTemplate.

Sjors commented:
While trying to start the Template Provider from `bitoin-mine` it crashes as soon as it tried any libsecp operation, at this line: `ARG_CHECK(secp256k1_ecmult_gen_context_is_built(&ctx->ecmult_gen_ctx));`

ryanofsky reviewed (commented):
re: https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2233301504

> If you split the capnp additions out of 4e1a434 it's a bit easier for me to combine this PR with #29432 (by omitting the interface changes commit here or there).

Hmm, what specifically would you like me to change here? This PR will be pretty difficult for me maintain if it based on other PRs, so it is just based on master.

Sjors commented:
> Hmm, what specifically would you like me to change here? 

Just to have the changes in `src/ipc/capnp` be their own commit. I need those in https://github.com/Sjors/bitcoin/pull/48, while at the same I need to avoid including the rest of 4e1a4342f3b2a857e3211587cd14e36704197483.

> The purpose of createNewBlock2 is just to make sure serialization works when createNewBlock is changed to return BlockTemplate.

In that case I'll just add a commit to https://github.com/Sjors/bitcoin/pull/48 that renames `createNewBlock2` to `createNewBlock`. That shouldn't be a problem to keep rebased.

ryanofsky commented:
Updated 4e1a4342f3b2a857e3211587cd14e36704197483 -> e1ae38f4bbcdb1b3fa5e1eab1abfa9cf2232798c ([`pr/mine.3`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.3) -> [`pr/mine.4`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.3..pr/mine.4)) adding several unit tests, and also starting to split up the "Add bitcoin-mine test program" commit by moving CTransaction serialization support to a new commit.

ryanofsky commented:
Updated e1ae38f4bbcdb1b3fa5e1eab1abfa9cf2232798c -> e1fa475a5becca0083f59d7cafac9bc67c2f4f1c ([`pr/mine.4`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.4) -> [`pr/mine.5`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.4..pr/mine.5)) adding support for -loglevel and other logging parameters as suggested https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2233469862

ryanofsky commented:
Updated e1fa475a5becca0083f59d7cafac9bc67c2f4f1c -> bdd68d5bca483071ca0729e6f0d2d71706ad21e7 ([`pr/mine.5`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.5) -> [`pr/mine.6`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.5..pr/mine.6)) fixing lint error and splitting up "Expand mining interface" commit into two parts.

re: https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2233601603

> Just to have the changes in `src/ipc/capnp` be their own commit. I need those in [Sjors#48](https://github.com/Sjors/bitcoin/pull/48), while at the same I need to avoid including the rest of [4e1a434](https://github.com/bitcoin/bitcoin/commit/4e1a4342f3b2a857e3211587cd14e36704197483).

Makes sense. The latest push splits up the "Expand mining interface" commit into two parts so it should be easier to cherry-pick the part you need now.

Sjors commented:
I ran into:

```
make -j17 src/bitcoin-mine
make: *** No rule to make target `src/bitcoin-mine'.  Stop.
```

Fixed with https://github.com/Sjors/bitcoin/commit/68979b8f4c4e069bc9b6d19679ef6832839ceae4

ryanofsky commented:
Rebased bdd68d5bca483071ca0729e6f0d2d71706ad21e7 -> fba04d7756c77ed9371c5dc90d3bebc9aa767f4b ([`pr/mine.6`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.6) -> [`pr/mine.7`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.6-rebase..pr/mine.7)) due to conflict with #30356

Sjors commented:
If you want to simplify this PR a bit further, you can drop `waitFeesChanged`, as well as `getCoinbaseMerklePath()` and `submitSolution` from the `BlockTemplate` interface.

I moved those out of the PRs on the repo and into my own fork (https://github.com/Sjors/bitcoin/pull/52 and https://github.com/Sjors/bitcoin/pull/53). I can apply the required changes to `ipc/capnp/mining.capnp` myself in https://github.com/Sjors/bitcoin/pull/48; that's much easier than I expected.

That should remove any friction from getting the mining interface changes from #30440 and #30409 merged.

We should then aim to get everything merged in the PR, maybe with the exception of `multiprocess: Add bitcoin-mine test program`.

By the time multiprocess is merged into Bitcoin Core (unlikely to make it into v28) I'll make PRs for the last missing bits. Those should be more fleshed out by then.

ryanofsky commented:
Rebased fba04d7756c77ed9371c5dc90d3bebc9aa767f4b -> ec5f45b9186184cac9c64ed08b8b8cc596e301a3 ([`pr/mine.7`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.7) -> [`pr/mine.8`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.7-rebase..pr/mine.8)), moving most code that was here into base PRs #30509 and #30510. I also followed up various review suggestions above like fixing the top-level make target and dropping unneeded parts of the mining interface.

Sjors commented:
Can you rebase this for CMake?

ryanofsky committed:
ipc: Add bitcoin-mine test program

See src/bitcoin-mine.cpp for usage information.

Co-authored-by: Sjors Provoost <sjors@sprovoost.nl>

ryanofsky commented:
Rebased ec5f45b9186184cac9c64ed08b8b8cc596e301a3 -> 697ba11c9496e3a2aaf6fc3ba64fd39a028109c2 ([`pr/mine.8`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.8) -> [`pr/mine.9`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.8-rebase..pr/mine.9)) just updating for cmake

ryanofsky commented:
Rebased 7b9aa0b6eb1126cb32da91f3c9d2df12325aa90c -> 2227afac0372358287fb879f3b0bd07fd653f3f8 ([`pr/mine.10`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.10) -> [`pr/mine.11`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.11), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.10-rebase..pr/mine.11)) after base PR's #30509 and #30510 were merged.

Marking ready for review

Sjors commented on src/bitcoin-mine.cpp:
```diff
@@ -0,0 +1,116 @@
+    if (tip) {
+        tfm::format(std::cout, "Tip hash is %s.\n", tip->hash.ToString());
+    } else {
+        tfm::format(std::cout, "Tip hash is null.\n");
```
I can't reach this condition. Even when I spin up a mainnet node and immediately start `bitcoin-mine` it just seems to wait a bit and then return a hash.

Maybe we should simplify the `getTip()` interface to not return an optional. Though I'm not sure if it's completely impossible to hit this.

Sjors commented:
tACK 2227afac0372358287fb879f3b0bd07fd653f3f8

Except I noticed that `bitcoin-mine -debug=ipc` doesn't print anything, which is surprising.

Tested the standalone application. I plan to do another rebase of https://github.com/Sjors/bitcoin/pull/48 on top of this later (which turns `bitcoin-mine` into a Template Provider).

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2385186175

Thanks for the review! Fixed the problem with `-debug=ipc` now. This was missing a call to parse `-debug` arguments.

Updated 2227afac0372358287fb879f3b0bd07fd653f3f8 -> 2f9c2f4c6dcf8c45514e99d1018301aba61940fc ([`pr/mine.11`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.11) -> [`pr/mine.12`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.12), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.11..pr/mine.12))

Sjors commented:
ACK 2f9c2f4c6dcf8c45514e99d1018301aba61940fc

Logging works now.

Sjors commented on src/bitcoin-mine.cpp:
```diff
@@ -0,0 +1,124 @@
+        tfm::format(std::cerr, "Error: Specified data directory \"%s\" does not exist.\n", args.GetArg("-datadir", ""));
+        return EXIT_FAILURE;
+    }
+    SelectParams(args.GetChainType());
```
Can you move this above `AddArgs`? In https://github.com/Sjors/bitcoin/pull/48 I'm adding some argument documentation that need accsess to `BaseParams`.

Sjors replied:
Never mind, this causes more headaches that it solves.

hebasto commented:
> In order to get `bitcoin-mine` in the guix build (once it supports multiprocess) I think you need to move `set(installable_targets)` up in the file, and also add `bitcoin-mine` to the list in `foreach(target IN ITEMS bitcoind` in `Maintenance.cmake`.

I don't think `set(installable_targets)`  needs to be moved up because it already precedes all subsequent use cases.

> Though I'm not sure why that doesn't use `installable_targets`

There are two local `installable_targets` variables: one scoped to the `src` directory and another to `src/qt`. This separation is necessary to be able to install GUI targets as part of the "GUI" component. On the other hand, the list of targets in the `add_maintenance_targets()` function is global and intended to hold targets built in the Guix script to run `symbol-check.py` and `security-check.py` on them.

Sjors commented:
@hebasto I forgot that I moved `target_link_libraries(bitcoin-mine` up myself in  https://github.com/Sjors/bitcoin/pull/48/commits/254cf216b5e312297ec34a4ee58256d21ab8571e. So need to change `set(installable_targets)` here.

Sjors commented on src/bitcoin-mine.cpp:
```diff
@@ -0,0 +1,124 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <config/bitcoin-config.h> // IWYU pragma: keep
```
2f9c2f4c6dcf8c45514e99d1018301aba61940fc: silent merge conflict with #30937.

Sjors commented on src/bitcoin-mine.cpp:
```diff
@@ -0,0 +1,124 @@
+        return EXIT_FAILURE;
+    }
+    if (HelpRequested(args) || args.IsArgSet("-version")) {
+        std::string output{strprintf("%s bitcoin-mine version", PACKAGE_NAME) + " " + FormatFullVersion() + "\n"};
```
`PACKAGE_NAME` changed in #31042

ryanofsky committed:
test: add interface_ipc_mining.py test calling bitcoin-mine

Co-authored-by: Sjors Provoost <sjors@sprovoost.nl>

ryanofsky commented:
Rebased 2f9c2f4c6dcf8c45514e99d1018301aba61940fc -> 1f18eefe1e5bc3a97f5f6c9637a3a193e2c2f22e ([`pr/mine.12`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.12) -> [`pr/mine.13`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.13), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.12-rebase..pr/mine.13)) fixing silent merge conflicts and adding a functional test and updating gen-manpages

---

re: https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2394194530
re: https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2396158697
re: https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2396197303

> In order to get `bitcoin-mine` in the guix build (once it supports multiprocess) I think you need to move `set(installable_targets)` up in the file, and also add `bitcoin-mine` to the list in `foreach(target IN ITEMS bitcoind` in `Maintenance.cmake`. Though I'm not sure why that doesn't use `installable_targets`, cc @hebasto.
>
> Maybe add a placeholder man page for `gen-manpages.py`.

Not sure I understood the outcome of this cmake discussion, but it seems like installable_targets and Maintenance.cmake changes might not be needed? Happy to add any changes here that would help.

Sjors commented:
I rebased https://github.com/Sjors/bitcoin/pull/48 to use the latest version of this PR, to figure out what I need for cmake. The only thing I still suggest is:

* add `bitcoin-mine` to the `foreach(target IN ITEMS` list in `cmake/module/Maintenance.cmake`. IIUC that will cause "the Guix script to run symbol-check.py and security-check.py on them". Even though this PR doesn't add bitcoin-mine to the guix build artefacts, it seems better to add it to this list now so we don't forget.

ACK 27072547bb22cdd2080d1014eba9c30bc9d47650 

I tested that `gen-manpages.py` generates something useful (`man doc/man/bitcoin-mine.1`).

The new functional test is nice.

Sjors commented:
re-utACK 99f2f335b6a8c2b5fe518a0c0056eb2a42c0496b

Sjors commented:
It would be nice to "Build bitcoin-mine executable." to the summary and add a similar guard as in #31834.

ryanofsky commented:
Rebased 99f2f335b6a8c2b5fe518a0c0056eb2a42c0496b -> 9194e461518e6b843c1a707ba1b10ab7a000e096 ([`pr/mine.15`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.15) -> [`pr/mine.16`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.16), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.15-rebase..pr/mine.16)) due to conflict with #31834. Also added status line for the executable as suggested

ryanofsky commented:
Rebased 59e1d4abb34453805a09a62c51ead7347787b2e3 -> b4f024ea336f72fd1023d0929761978566126740 ([`pr/mine.17`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.17) -> [`pr/mine.18`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.18), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.17-rebase..pr/mine.18)) due to conflict with #31865

Sjors reviewed (commented):
Code reviewed 5554f9cb4abf7e1852dd79d2d75e714fa8cbeadd.

Rebase looks good, but found a bug in the test.

ryanofsky commented on test/functional/interface_ipc_mining.py:
```diff
@@ -0,0 +1,43 @@
+        self.binary_paths.bitcoind = self.binary_paths.bitcoin_node
+
+        # Work around default CI path exceeding maximum socket path length
+        if len(self.options.tmpdir + "/node0/regtest/node.sock") < 108:
```
re: https://github.com/bitcoin/bitcoin/pull/30437#discussion_r2007217890

Thanks! Applied fix

ryanofsky reviewed (commented):
Updated 5554f9cb4abf7e1852dd79d2d75e714fa8cbeadd -> 0ec78961915da141b9c68a39ec0ebd5091c13be0 ([`pr/mine.20`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.20) -> [`pr/mine.21`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.21), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.20..pr/mine.21)) adding suggested test fix

Sjors commented:
ACK 0ec78961915da141b9c68a39ec0ebd5091c13be0

ryanofsky commented:
Updated 0ec78961915da141b9c68a39ec0ebd5091c13be0 -> fae300f159cd25a12abf4d5fbb93135cececd38d ([`pr/mine.21`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.21) -> [`pr/mine.22`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.22), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.21..pr/mine.22)) just tweaking comments

Sjors commented:
re-ACK fae300f159cd25a12abf4d5fbb93135cececd38d

ismaelsadeeq reviewed (commented):
Concept ACK

This can serve as a usage example for the interfaces.

I've rebased this on `master`: https://github.com/ismaelsadeeq/bitcoin/tree/05-2025-bitcoin-mine-rebase and did a few tests.

1 - Ran `bitcoin-mine` without a `bitcoin-node` process — it executed correctly and showed an appropriate error message.
<details>
  <summary>logs</summary>
  
  ```terminal
2025-05-19T21:00:18Z Default data directory /Users/abubakarismail/Library/Application Support/Bitcoin
2025-05-19T21:00:18Z Using data directory /Users/abubakarismail/Library/Application Support/Bitcoin/regtest
2025-05-19T21:00:18Z Config file: /Users/abubakarismail/Library/Application Support/Bitcoin/bitcoin.conf (not found, skipping)
2025-05-19T21:00:18Z Command-line arg: debug=""
2025-05-19T21:00:18Z Command-line arg: regtest=""
Error: Connection refused
Probably bitcoin-node is not running or not listening on a unix socket. Can be started with:

    bitcoin-node -chain=regtest -ipcbind=unix
  ```
  </details>
2 - Ran `bitcoin-mine` with a `bitcoin-node` process running, but without the `-ipcbind` option.
  <details>
  <summary>logs</summary>
  
  ```terminal
2025-05-19T21:00:18Z Default data directory /Users/abubakarismail/Library/Application Support/Bitcoin
2025-05-19T21:00:18Z Using data directory /Users/abubakarismail/Library/Application Support/Bitcoin/regtest
2025-05-19T21:00:18Z Config file: /Users/abubakarismail/Library/Application Support/Bitcoin/bitcoin.conf (not found, skipping)
2025-05-19T21:00:18Z Command-line arg: debug=""
2025-05-19T21:00:18Z Command-line arg: regtest=""
Error: Connection refused
Probably bitcoin-node is not running or not listening on a unix socket. Can be started with:

    bitcoin-node -chain=regtest -ipcbind=unix
  ```
  </details>
3 - Ran `bitcoin-mine` with a `bitcoin-node` process and the `-ipcbind` option set.

  <details>
  <summary>logs</summary>
  
  ```terminal
2025-05-19T16:09:06Z Default data directory /Users/abubakarismail/Library/Application Support/Bitcoin
2025-05-19T16:09:06Z Using data directory /Users/abubakarismail/Library/Application Support/Bitcoin/regtest
2025-05-19T16:09:06Z Config file: /Users/abubakarismail/Library/Application Support/Bitcoin/bitcoin.conf (not found, skipping)
2025-05-19T16:09:06Z Command-line arg: debug=""
2025-05-19T16:09:06Z Command-line arg: regtest=""
2025-05-19T16:09:06Z [ipc] {bitcoin-mine-85594/19388374} IPC client first request from current thread, constructing waiter
2025-05-19T16:09:06Z [ipc] {bitcoin-mine-85594/19388374} IPC client send Init.construct$Params (threadMap = <external capability>)
2025-05-19T16:09:06Z [ipc] {bitcoin-mine-85594/19388374} IPC client recv Init.construct$Results (threadMap = <external capability>)
Connected to bitcoin-node
2025-05-19T16:09:06Z [ipc] {bitcoin-mine-85594/19388374} IPC client send Init.makeMining$Params (context = (thread = <external capability>, callbackThread = <external capability>))
2025-05-19T16:09:06Z [ipc] {bitcoin-mine-85594/19388374} IPC client recv Init.makeMining$Results (result = <external capability>)
2025-05-19T16:09:06Z [ipc] {bitcoin-mine-85594/19388374} IPC client send Mining.getTip$Params (context = (thread = <external capability>, callbackThread = <external capability>))
2025-05-19T16:09:06Z [ipc] {bitcoin-mine-85594/19388374} IPC client recv Mining.getTip$Results (result = (hash = "\\311\\353\\275\\006z\\366\\213\\b\\266*<\\226\\363\\250\\345\\235\\214E\\203\\300\\2323YI+\\233=5\\365\\222\\001J", height = 2012), hasResult = true)
Tip hash is 4a0192f5353d9b2b4959339ac083458c9de5a8f3963c2ab6088bf67a06bdebc9.
2025-05-19T16:09:06Z [ipc] {bitcoin-mine-85594/19388374} IPC client destroy N2mp11ProxyClientIN3ipc5capnp8messages6MiningEEE
2025-05-19T16:09:06Z [ipc] {bitcoin-mine-85594/19388374} IPC client destroy N2mp11ProxyClientIN3ipc5capnp8messages4InitEEE
2025-05-19T16:09:06Z [ipc] {bitcoin-mine-85594/b-capnp-loop-19388377} EventLoop::loop done, cancelling event listeners.
2025-05-19T16:09:06Z [ipc] {bitcoin-mine-85594/b-capnp-loop-19388377} EventLoop::loop bye.
  ```
  </details>

ryanofsky commented on src/init/bitcoin-mine.cpp:
```diff
@@ -0,0 +1,29 @@
+// Copyright (c) 2025 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
```
re: https://github.com/bitcoin/bitcoin/pull/30437#discussion_r2096468681

Good suggestion, this is now moved to an earlier commit, and it is now sharing code with #32297 (MakeMineInit is now replaced by MakeBasicInit originally introduced in that PR)

ismaelsadeeq replied:
this looks much better, I will review again soon.

ryanofsky commented on src/bitcoin-mine.cpp:
```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2025 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
```
re: https://github.com/bitcoin/bitcoin/pull/30437#discussion_r2096479030

Yes I think depending on what uses there may for this in the future I think it could be renamed to something more generic. For example I could see this potentially being used to test more IPC functionality from functional tests, and if that would happen a name like `bitcoin-ipc` or `bitcoin-test-ipc` could make more sense. I think for now the name `bitcoin-mine` is reasonable since it's only connecting and using the mining interface, and it was originally added to start implementing external stratum support. Did not rename for now, but open to more suggestions about whether/how to rename

ismaelsadeeq replied:
+1 `bitcoin-test-ipc` sounds good to me

Sjors replied:
I would also prefer to keep this binary focused on testing / demoing the mining interface.

ismaelsadeeq replied:
cc @TheCharlatan  since you wrote https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2984042087

Just resurfacing a previous comment along with a recent one.

I think this should remain a simple IPC boilerplate. You mentioned “it's quite a bit” perhaps  you should indicate what you feel is unnecessary and suggest ways we could slim it down.

@Sjors:

> Binary focused on testing / demoing the mining interface.

Don’t we already have unit tests for the interface? If the goal is to demo the mining interface, we may end up implementing more mining-related utilities, which increases the scope unnecessarily especially with the recent updates.

I think the IPC demo should focus on showing how to create an IPC client, which it already does well.
The mining interface tests already confirm that the methods work as expected. Reimplementing and re-testing them via this binary is redundant I think.

ryanofsky reviewed (commented):
Rebased fae300f159cd25a12abf4d5fbb93135cececd38d -> 6645095e6ae0d9c208a47fc7322d859c4fe6afb0 ([`pr/mine.22`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.22) -> [`pr/mine.23`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.23), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.22-rebase..pr/mine.23)) due to conflicts with #28710 and also pulled in MakeBasicInit function to share code with #32297

ismaelsadeeq commented:
Not sure were to report, so I will drop this here.

While writing a custom ipc-client, I noticed that the IPC will freeze when you shutdown the node while the ipc-client is still running.

How to reproduce

- Apply the following patch to this branch:
<details>
<summary>diff</summary>

```diff
diff --git a/src/bitcoin-mine.cpp b/src/bitcoin-mine.cpp
index e31f666a68e..3b9e1b81964 100644
--- a/src/bitcoin-mine.cpp
+++ b/src/bitcoin-mine.cpp
@@ -16,6 +16,11 @@
 #include <logging.h>
 #include <tinyformat.h>
 #include <util/translation.h>
+#include <util/time.h>
+
+#ifndef WIN32
+#include <signal.h>
+#endif
 
 static const char* const HELP_USAGE{R"(
 bitcoin-mine is a test program for interacting with bitcoin-node via IPC.
@@ -48,6 +53,24 @@ static void AddArgs(ArgsManager& args)
     init::AddLoggingArgs(args);
 }
 
+#ifndef WIN32
+static bool g_interrupt{false};
+
+static void registerSignalHandler(int signal, void(*handler)(int))
+{
+    struct sigaction sa;
+    sa.sa_handler = handler;
+    sigemptyset(&sa.sa_mask);
+    sa.sa_flags = 0;
+    sigaction(signal, &sa, nullptr);
+}
+static void HandleSIGTERM(int)
+{
+    g_interrupt = true;
+}
+#endif
+
+
 MAIN_FUNCTION
 {
     ArgsManager& args = gArgs;
@@ -112,7 +135,6 @@ MAIN_FUNCTION
     tfm::format(std::cout, "Connected to bitcoin-node\n");
     std::unique_ptr<interfaces::Mining> mining{node_init->makeMining()};
     assert(mining);
-
     auto tip{mining->getTip()};
     if (tip) {
         tfm::format(std::cout, "Tip hash is %s.\n", tip->hash.ToString());
@@ -120,5 +142,14 @@ MAIN_FUNCTION
         tfm::format(std::cout, "Tip hash is null.\n");
     }
 
+#ifndef WIN32
+    registerSignalHandler(SIGTERM, HandleSIGTERM);
+    registerSignalHandler(SIGINT, HandleSIGTERM);
+#endif
+
+    while(!g_interrupt) {
+        UninterruptibleSleep(1000ms);
+    }
+
     return EXIT_SUCCESS;
 }
```

</details

- Start the bitcoin-node process:

```terminal
build/bin/bitcoin-node -chain=regtest -ipcbind=unix -debug
```

- Start the client

```terminal
build/bin/bitcoin-mine -chain=regtest -debug
=1
```


- Stop the node while client is still running, After the node shuts down, the IPC becomes unresponsive and hangs.

```terminal
build/bin/bitcoin-cli -regtest stop

....
2025-05-28T17:11:42Z [bench] FlushStateToDisk: write block index to disk completed (2.98ms)
2025-05-28T17:11:42Z [bench] FlushStateToDisk: write coins cache to disk (0 coins, 256KiB) started
2025-05-28T17:11:42Z [coindb] Writing final batch of 0.00 MiB
2025-05-28T17:11:42Z [leveldb] WriteBatch memory usage: db=chainstate, before=0.0MiB, after=0.0MiB
2025-05-28T17:11:42Z [coindb] Committed 0 changed transaction outputs (out of 0) to coin database...
2025-05-28T17:11:42Z [bench] FlushStateToDisk: write coins cache to disk (0 coins, 256KiB) completed (0.04ms)
2025-05-28T17:11:42Z [validation] Enqueuing ChainStateFlushed: block hash=4a0192f5353d9b2b4959339ac083458c9de5a8f3963c2ab6088bf67a06bdebc9
2025-05-28T17:11:42Z Shutdown: done
```

- After shutting down the client, the ipc will also shutdown
```
2025-05-28T17:12:19Z [ipc] {bitcoin-node-82440/b-capnp-loop-24295893} IPC server destroy N2mp11ProxyServerIN3ipc5capnp8messages6MiningEEE
2025-05-28T17:12:19Z [ipc] {bitcoin-node-82440/b-capnp-loop-24295893} IPC server: socket disconnected.
2025-05-28T17:12:19Z [ipc] {bitcoin-node-82440/b-capnp-loop-24295893} IPC server destroy N2mp11ProxyServerIN3ipc5capnp8messages4InitEEE
2025-05-28T17:12:19Z [ipc] {bitcoin-node-82440/b-capnp-loop-24295893} EventLoop::loop done, cancelling event listeners.
2025-05-28T17:12:19Z [ipc] {bitcoin-node-82440se #3 Mining.getTip$Results (result
```


I can also reproduce this on https://github.com/Sjors/bitcoin/pull/48

Sjors commented on src/bitcoin-mine.cpp:
```diff
@@ -0,0 +1,124 @@
+        return EXIT_FAILURE;
+    }
+
+    // Connect to existing bitcoin-node process or spawn new one.
```
IIRC we decided to not include spawning support just yet to keep it simple, so the comment may be outdated.

ryanofsky reviewed (commented):
Thanks for the reviews!

Rebased 6645095e6ae0d9c208a47fc7322d859c4fe6afb0 -> 27c6576aba5c59402b8844703e04face2f41578c ([`pr/mine.23`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.23) -> [`pr/mine.24`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.24), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.23-rebase..pr/mine.24)) due to conflict with #31375, also bringing over some changes from #32297 and implementing some suggestions.

re: https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2917116354

> Not sure were to report, so I will drop this here.

Thanks! This is a known issue fixed by #32345 (specifically 8ca7049ceac3ce28961aeb7084a6e82e43a8752c).

re: https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2918504989

> When rebasing, you may also need to add a resource file like in #32634. Although we don't ship this binary for Windows builds yet, so maybe that can wait.

Thanks! Added now.

Sjors commented:
Now that #31375 landed, can you add `bitcoin mine` as a command?

Other than that 27c6576aba5c59402b8844703e04face2f41578c looks good and works.

ryanofsky committed:
ipc: Add bitcoin-mine to bitcoin wrapper interface

Suggested by Sjors Provoost <sjors@sprovoost.nl> in
https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2939117270

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2939117270

> Now that #31375 landed, can you add `bitcoin mine` as a command?

Oops, somehow i thought I had already done that..

Added 1 commit 27c6576aba5c59402b8844703e04face2f41578c -> 278101f6ee723e6401ddb3e46f2fce3b60cc42ca ([`pr/mine.24`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.24) -> [`pr/mine.25`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.25), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.24...pr/mine.25)) to support `bitcoin mine` command

TheCharlatan commented:
I am a bit confused what this binary is supposed to be good for. For a pure documentation binary it has quite a bit of boilerplate, but also does not show how to actually use the interface. Why not make this a utility that attempts to mine a block in `n` tries or something similar? Maybe something close to generateblock (though that would require adding some basic listeners for notifications), or something like (which was good enough to mine some orphan testnet4 blocks):

```diff
diff --git a/src/bitcoin-mine.cpp b/src/bitcoin-mine.cpp
index 07e221d11e..e124a9f916 100644
--- a/src/bitcoin-mine.cpp
+++ b/src/bitcoin-mine.cpp
@@ -9,0 +10 @@
+#include <consensus/merkle.h>
@@ -12,0 +14 @@
+#include <cstdlib>
@@ -15,0 +18 @@
+#include <key_io.h>
@@ -16,0 +20 @@
+#include <pow.h>
@@ -39,0 +44 @@ const TranslateFn G_TRANSLATION_FUN{nullptr};
+const uint64_t DEFAULT_MAX_TRIES{10'000};
@@ -47,0 +53 @@ static void AddArgs(ArgsManager& args)
+    args.AddArg("-maxtries=<n>", strprintf("Try to mine a block for <n> tries. Default %d", DEFAULT_MAX_TRIES), ArgsManager::ALLOW_ANY, OptionsCategory::OPTIONS);
@@ -120,0 +127,21 @@ MAIN_FUNCTION
+        return EXIT_SUCCESS;
+    }
+
+    auto consensus_params{Params().GetConsensus()};
+
+    uint64_t max_tries{std::max<uint64_t>(DEFAULT_MAX_TRIES, args.GetIntArg("-maxtries", DEFAULT_MAX_TRIES))};
+    auto tries_remaining{max_tries};
+    auto block_template{mining->createNewBlock({})};
+    auto block{block_template->getBlock()};
+    block.hashMerkleRoot = BlockMerkleRoot(block);
+
+    while (tries_remaining> 0 && block.nNonce < std::numeric_limits<uint32_t>::max() && !CheckProofOfWork(block.GetHash(), block.nBits, consensus_params)) {
+        ++block.nNonce;
+        --tries_remaining;
+    }
+    block_template->submitSolution(block.nVersion, block.nTime, block.nNonce, block_template->getCoinbaseTx());
+
+    if (tip->hash != mining->getTip()->hash) {
+        tfm::format(std::cout, "Mined a block, tip advanced to %s.\n", mining->getTip()->hash.ToString());
+    } else {
+        tfm::format(std::cout, "Failed to mine a block in %d iterations. Try again. \n", max_tries);
```

sedited committed:
bitcoin-mine: Extend example to mine a block

This includes createNewBlock fix from #34860 to allow mining blocks at
heights <16.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

Sjors commented:
@TheCharlatan that should definitely be enough for regtest blocks, which would have the benefit of (almost certain) `submitSolution()` coverage.

ryanofsky reviewed (commented):
Rebased 278101f6ee723e6401ddb3e46f2fce3b60cc42ca -> 1a87ed000e09b10c173ba87d4bf2f79bf63abfa5 ([`pr/mine.25`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.25) -> [`pr/mine.26`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.26), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.25-rebase..pr/mine.26)) to resolve conflicts. Also split commits to separate code shared with #32297 and applied suggestions on test environment variables and extending the mining example to generate a block.

---

re: https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-2984042087

> I am a bit confused what this binary is supposed to be good for. For a pure documentation binary it has quite a bit of boilerplate, but also does not show how to actually use the interface.

I think documenting the boilerplate is basically of the point of the PR. The interface itself is just a normal C++ object that should be straightfoward to use. But getting a pointer to the interface is less straightforward. You start with a client `Init` object, call `init->ipc()->connectAddress()` on that to get a pointer to a server `Init` object, and call `makeMining()` on that to get a pointer to the mining interface. If the client is a standalone program like `bitcoin-mine` it also needs to deal with options parsing, logging, connection errors, and exception handling. I think seeing this is useful for understanding the IPC mechanism and also useful as a starting point for writing new IPC clients.

I do think the submitSolution example you posted is helpful too, so added that as a new commit.

Sjors commented:
Linter is unhappy over an `unused import` in the test.

ryanofsky commented:
Updated 1a87ed000e09b10c173ba87d4bf2f79bf63abfa5 -> 2910315d1d8a3879f4e0bd50bf90870b642bfede ([`pr/mine.26`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.26) -> [`pr/mine.27`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.27), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.26..pr/mine.27)) fixing unused python import lint error https://cirrus-ci.com/task/5421333066022912

Sjors commented:
ACK 2910315d1d8a3879f4e0bd50bf90870b642bfede

ismaelsadeeq reviewed (commented):
Weak ACK 2910315d1d8a3879f4e0bd50bf90870b642bfede

I’m not sure whether 600a52a0ba08dca06de891e6716ab2730580ddcd is necessary.

Personally, I preferred the previous iteration keeping this PR as a minimal boilerplate to demonstrate IPC usage. This increases the scope quite a bit, and if we're going to demonstrate mining, I think it should be done accurately.

I’ve added some comments below pointing out that we don’t currently handle chain tip changes. Could also be better to use `waitNext` and be `regtest` only.

Sjors commented:
> This increases the scope quite a bit, and if we're going to demonstrate mining, I think it should be done accurately.

I don't think the goal is to _demonstrate_ mining, it's to increase (functional) test coverage.

Sjors commented on src/bitcoin-mine.cpp:
```diff
@@ -120,5 +127,25 @@ MAIN_FUNCTION
+        ++block.nNonce;
+        --tries_remaining;
+    }
+    block_template->submitSolution(block.nVersion, block.nTime, block.nNonce, block_template->getCoinbaseTx());
```
For the purpose of test coverage, it's good to call it this with an invalid solution.

Sjors commented on test/functional/interface_ipc_mining.py:
```diff
@@ -48,7 +46,9 @@ def setup_nodes(self):
         args = [self.binary_paths.bitcoinmine, f"-datadir={self.nodes[0].datadir_path}"] + self.mine_args
         result = subprocess.run(args, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True, check=True)
-        assert_equal(result.stdout, "Connected to bitcoin-node\nTip hash is 0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206.\n")
+        expect = r"Connected to bitcoin-node\nTip hash is 0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206\.\n(Mined a block|Failed to mine a block).*?\n"
```
Regtest has an extremely low difficulty, so it's (almost?) impossible to not find a solution. However it does seem more correct to handle the failure case.

ryanofsky commented on src/bitcoin-mine.cpp:
```diff
@@ -120,5 +127,25 @@ MAIN_FUNCTION
+
+    while (tries_remaining> 0 && block.nNonce < std::numeric_limits<uint32_t>::max() && !CheckProofOfWork(block.GetHash(), block.nBits, consensus_params)) {
+        ++block.nNonce;
+        --tries_remaining;
```
re: https://github.com/bitcoin/bitcoin/pull/30437#discussion_r2213947741

> This should also break when the chain tip changes or generate a new template. Right now, it's possible to pass a large `max_tries` and get stuck.

I agree this is not ideal, though I don't think it's necessarily a problem. Probably more ideally this code would use two threads and increment the nonce in one thread and call waitNext in another thread, updating the hash if the tip changes. Calling getTip() every time the nonce is incremented would slow down the miner by orders of magnitude. (Even with multiple threads it's not clear that checking the tip hash frequently instead of only after a very large number of tries would be an improvement.) For now just added a comment about this.

ryanofsky reviewed (commented):
Thanks for the reviews!

Updated 2910315d1d8a3879f4e0bd50bf90870b642bfede -> 9d8851c0103f0dff8d659950e48d70ec86c86b30 ([`pr/mine.27`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.27) -> [`pr/mine.28`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.28), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.27..pr/mine.28)) just updating two code comments.
Updated 9d8851c0103f0dff8d659950e48d70ec86c86b30 -> 3d099384a60c8f64c8d67a9f1e7b8649a9c54313 ([`pr/mine.28`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.28) -> [`pr/mine.29`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.29), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.28..pr/mine.29)) fixing typo in a new comment.

---

re: https://github.com/bitcoin/bitcoin/pull/30437#pullrequestreview-3029237466

@ismaelsadeeq I'm actually not sure I understand the source of your concerns about the mining example (commit 2fbe6ea8557b21be9e48839500d3177abdce18e0). Even if it is not a perfect example, I feel like it's a pretty informative demonstration of how the mining works and it provides some cross-process test coverage of mining interface createNewBlock, getBlock, and submitSolution methods. While we do have existing unit test coverage for these methods, the unit tests bypass the socket i/o and serialization code used in bitcoin-mine. So the end-to-end coverage provided here seems useful, since there is not much similar coverage.

If commit 2fbe6ea8557b21be9e48839500d3177abdce18e0 is harmful in some way, I think it's nonessential and would be fine to drop. But I don't really see downsides to including it.

Sjors commented:
re-utACK 3d099384a60c8f64c8d67a9f1e7b8649a9c54313

ismaelsadeeq commented:
re-ACK e1f139bd5fc9ee737d2b08307fca2b33354c5747

> I feel like adding more special cases to this code will make it less less useful as a demonstration of what the mining API looks like and how it can be called. 

Previously, I assumed that the purpose of this commit was to demonstrate how the mining API works. @Sjors highlighted it's for test coverage here https://github.com/bitcoin/bitcoin/pull/30437#issuecomment-3087403014. Your recent comment seems to reinforce my original assumption.

> I'm open to adding more checks if it will solve a problem though (I just need a better understanding of what the potential problem is)

Definitely not a problem I was just thinking If it's for testing purposes, we perform tests in `-regtest` mode, both in functional and unit tests. Hence should be restricted to `regtest`.
Although the likelihood of finding a valid solution using a CPU is nearly zero, we still shouldn’t provide the avenue for it .

Sjors commented:
re-ACK e1f139bd5fc9ee737d2b08307fca2b33354c5747

BrandonOdiwuor reviewed (commented):
https://github.com/bitcoin/bitcoin/pull/30437/files ipc: add bitcoin-mine test program #30437
Tested ACK e1f139bd5fc9ee737d2b08307fca2b33354c5747

I tested the new `bitcoin-mine` IPC test program and `bitcoin -m mine` wrapper against a running `bitcoin-node` on macOS 15.0 in -regtest mode, verifying successful IPC connection, tip hash retrieval, and valid block mining/submission without errors

<img width="1512" height="918" alt="Screenshot 2025-08-12 at 13 16 19" src="https://github.com/user-attachments/assets/f71bd833-38a8-4a87-baee-39aa52be8252" />

ryanofsky reviewed (commented):
<!-- begin push-35 -->
Updated 2cb5e6472c8c0109390258c298915dd4011f0292 -> 057ea0a677293b6ecaa48008228ca3086ab9e0ff ([`pr/mine.34`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.34) -> [`pr/mine.35`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.35), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.34..pr/mine.35))<!-- end --> just tweaking bitcoin-mine status message
<!-- begin push-36 -->
Rebased 057ea0a677293b6ecaa48008228ca3086ab9e0ff -> 69d23a03ef962f889e1c100782199dac5cd3f600 ([`pr/mine.35`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.35) -> [`pr/mine.36`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.36), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.35-rebase..pr/mine.36))<!-- end --> due to conflicts with #34452
<!-- begin push-37 -->
Rebased 69d23a03ef962f889e1c100782199dac5cd3f600 -> bffde2f1ff8341a0100db50dd16bd4fd5d18f6c7 ([`pr/mine.36`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.36) -> [`pr/mine.37`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.37), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.36-rebase..pr/mine.37))<!-- end --> due to conflict with #34860
<!-- begin push-38 -->
Rebased bffde2f1ff8341a0100db50dd16bd4fd5d18f6c7 -> 5206449b9eec1ac61f890d2482a5e59a643dddd1 ([`pr/mine.37`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.37) -> [`pr/mine.38`](https://github.com/ryanofsky/bitcoin/commits/pr/mine.38), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine.37-rebase..pr/mine.38))<!-- end --> updating bitcoin-mine.cpp for new APIs and null check, due to conflicts with #33633 and #35551

Sjors commented:
ACK 057ea0a677293b6ecaa48008228ca3086ab9e0ff

Sjors commented on CMakeLists.txt:
```diff
@@ -668,6 +668,7 @@ message("  bitcoin-tx .......................... ${BUILD_TX}")
 message("  bitcoin-util ........................ ${BUILD_UTIL}")
 message("  bitcoin-wallet ...................... ${BUILD_WALLET_TOOL}")
 message("  bitcoin-chainstate (experimental) ... ${BUILD_UTIL_CHAINSTATE}")
+message("  bitcoin-mine (experimental) ......... ${bitcoin_daemon_status}")
```
It seems similar to `bitcoin-chainstate`: a demo and test app that we should at least build in dev preset, but probably don't need to ship in a release.

Perhaps it's better to move both of them to an "examples" folder?

Regarding Windows: support is still on the agenda for libmultiprocess: https://github.com/bitcoin-core/libmultiprocess/pull/231

And specifically for mining, my understanding is that people actually use that operating system. Although at the same time so far I haven't had anyone ask for it for sv2-tp (https://github.com/stratum-mining/sv2-tp/issues/7).

w0xlt commented on src/bitcoin-mine.cpp:
```diff
@@ -0,0 +1,158 @@
+
+    uint64_t max_tries{std::max<uint64_t>(DEFAULT_MAX_TRIES, args.GetIntArg("-maxtries", DEFAULT_MAX_TRIES))};
+    auto tries_remaining{max_tries};
+    auto block_template{mining->createNewBlock({}, /*cooldown=*/false)};
```
`Mining::createNewBlock()` can legally return `nullptr` when bitcoin-node is shutting down or mining is interrupted.

```diff
diff --git a/src/bitcoin-mine.cpp b/src/bitcoin-mine.cpp
index f02ed3dfba..d33c9e2326 100644
--- a/src/bitcoin-mine.cpp
+++ b/src/bitcoin-mine.cpp
@@ -133,6 +133,10 @@ MAIN_FUNCTION
     uint64_t max_tries{std::max<uint64_t>(DEFAULT_MAX_TRIES, args.GetIntArg("-maxtries", DEFAULT_MAX_TRIES))};
     auto tries_remaining{max_tries};
     auto block_template{mining->createNewBlock({}, /*cooldown=*/false)};
+    if (!block_template) {
+        tfm::format(std::cerr, "Error: Failed to create block template. bitcoin-node may be shutting down.\n");
+        return EXIT_FAILURE;
+    }
     auto block{block_template->getBlock()};
     block.hashMerkleRoot = BlockMerkleRoot(block);
```

w0xlt commented on src/bitcoin-mine.cpp:
```diff
@@ -0,0 +1,158 @@
+
+    auto consensus_params{Params().GetConsensus()};
+
+    uint64_t max_tries{std::max<uint64_t>(DEFAULT_MAX_TRIES, args.GetIntArg("-maxtries", DEFAULT_MAX_TRIES))};
```
With `std::max(DEFAULT_MAX_TRIES, ...)`, `max_tries` can never be lower than `DEFAULT_MAX_TRIES`, so values like `-maxtries=0` or `-maxtries=100` are ignored and become `10000`. Is this intentional ?

Also, `GetIntArg()` returns a signed integer, but `std::max<uint64_t>()` converts it to unsigned. A negative value like `-maxtries=-1` can become a huge `uint64_t`.

```diff
diff --git a/src/bitcoin-mine.cpp b/src/bitcoin-mine.cpp
index f02ed3dfba..d92a45501e 100644
--- a/src/bitcoin-mine.cpp
+++ b/src/bitcoin-mine.cpp
@@ -130,9 +130,18 @@ MAIN_FUNCTION
 
     auto consensus_params{Params().GetConsensus()};
 
-    uint64_t max_tries{std::max<uint64_t>(DEFAULT_MAX_TRIES, args.GetIntArg("-maxtries", DEFAULT_MAX_TRIES))};
+    int64_t max_tries_arg{args.GetIntArg("-maxtries", DEFAULT_MAX_TRIES)};
+    if (max_tries_arg < 0) {
+        tfm::format(std::cerr, "Error: -maxtries must be non-negative.\n");
+        return EXIT_FAILURE;
+    }
+    uint64_t max_tries{static_cast<uint64_t>(max_tries_arg)};
     auto tries_remaining{max_tries};
     auto block_template{mining->createNewBlock({}, /*cooldown=*/false)};
 ```

w0xlt reviewed (commented):
A few review comment:

Sjors commented:
It might be good to emphasize in the PR description that this adds coverage for how a libmultiprocess based client interacts with the node. Our functional tests do not use libmutltiprocess on the client side. We know from experience with sv2-tp vs. the Rust based SRI implementation that these differences can be important.

> Exposing this as `bitcoin-mine` even if installed in libexec and not in binary releases could make it seem like it has more utility than it has, or is ever likely to have.

Yes, I think this should be seen as a combination of a usage example and test coverage.
